benchmark-base/src/main/java/org/conscrypt/CipherEncodingBenchmark.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (C) 2017 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.conscrypt;
+
+import java.nio.ByteBuffer;
+import java.security.Key;
+import javax.crypto.Cipher;
+import javax.crypto.KeyGenerator;
+
+/**
+ * Benchmark for comparing cipher encoding performance.
+ */
+public final class CipherEncodingBenchmark {
+    private static final String ALGORITHM = "AES";
+    private static final int KEY_SIZE = 128;
+
+    public enum BufferType { ARRAY, HEAP, DIRECT }
```

Comment:
nit: IIRC ther eis some code that behaves differently if the source and destination are heap vs non heap.  For the byte buffer, you might consider having separate source and destination HEAP_DIRECT, HEAP_HEAP, DIRECT_HEAP, and DIRECT_DIRECT cases.

Reply:
That would really blow up the benchmark matrix ... not sure it's worth it.

Reply:
If you look at the CipherSpi.engineDoFinal, you can see they have 3 disctinct cases based on if the buffer is array backed or not.  I think it would be worth the extra cases, since the existing implementation treats each one differently.

Reply:
Done ... I've updated the benchmark results. Nothing too unexpected.